### PR TITLE
[#2267] Add AMQP 1.0 based DeviceConnectionClient implementation

### PR DESCRIPTION
**Pull request by sophokles73**
Added transport protocol agnostic DeviceConnectionClient interface to
the adapter client module.
Added AMQP 1.0 based implementation which simply wraps the existing
vertx-proton based "legacy" DeviceConnectionClientImpl.

### CODE REVIEW

File: adapters/pom.xml [hunk 1]
```diff
@@ -113,6 +113,10 @@
       <groupId>org.eclipse.hono</groupId>
       <artifactId>service-base-test-utils</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.eclipse.hono</groupId>
+      <artifactId>hono-service-base</artifactId>
+    </dependency>
```
Comment (calohmn) [inline adapters/pom.xml:119]: Corresponding dependency can be removed from the dependent "hono-adapter-base-spring" module then.

File: services/device-connection/pom.xml [hunk 2]
```diff
@@ -25,6 +25,16 @@
       <groupId>org.eclipse.hono</groupId>
       <artifactId>client-device-connection-infinispan</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.eclipse.hono</groupId>
+      <artifactId>hono-service-base</artifactId>
+      <exclusions>
+        <exclusion>
+          <groupId>org.eclipse.hono</groupId>
+          <artifactId>hono-client-adapter-amqp</artifactId>
```
Comment (calohmn) [inline services/device-connection/pom.xml:34]: Wouldn't it make sense to remove the 'hono-service-base' dependency here and add the exclusion to the existing 'hono-service-base' in the parent 'hono-services' module?

File: clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/command/amqp/ProtonBasedDeviceConnectionClient.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+ */
+public class ProtonBasedDeviceConnectionClient extends AbstractRequestResponseClient<DeviceConnectionResult>
+        implements DeviceConnectionClient {
+
+    private final CachingClientFactory<org.eclipse.hono.client.DeviceConnectionClient> clientFactory;
+
+    /**
+     * Creates a new client for a connection.
+     *
+     * @param connection The connection to the Credentials service.
```
Comment (calohmn) [inline clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/command/amqp/ProtonBasedDeviceConnectionClient.java:52]: Device Connection service

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java [hunk 4]
```diff
@@ -1010,9 +993,9 @@ protected void onCommandConnectionEstablished(final HonoConnection commandConnec
                 .map(client -> client.isConnected())
                 .orElseGet(() -> Future.failedFuture(new ServerErrorException(
                         HttpURLConnection.HTTP_UNAVAILABLE, "Command & Control client factory is not set"))));
-        connections.add(Optional.ofNullable(deviceConnectionClientFactory)
+        connections.add(Optional.ofNullable(deviceConnectionClient)
                 .map(client -> {
-                    if (deviceConnectionClientFactory instanceof ConnectionLifecycle) {
+                    if (deviceConnectionClient instanceof ConnectionLifecycle) {
```
[the remark this answers is not in the document]
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java:998]: yes, I forgot about that. Great suggestion! I will remove the *isConnected* method ...

File: services/device-connection/pom.xml (hunk 2 above)
Comment (sophokles73) [reply services/device-connection/pom.xml:34]: absolutely

**Comment by sophokles73** [comment]
@calohmn I have pushed changes ...

**Review by calohmn — APPROVED** [review]
LGTM

**Pull request merged**